src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -497,34 +497,75 @@ int encrypt(byte[] in, int inOfs, int len) {
      * @param len the length of the input data
      * @param out the buffer for the result
      * @param outOfs the offset in <code>out</code>
-     * @exception ProviderException if <code>len</code> is not
-     * a multiple of the block size
      * @return the number of bytes placed into the <code>out</code> buffer
      */
-    int encrypt(byte[] in, int inOfs, int len, byte[] out, int outOfs) {
-        checkDataLength(processed, len);
+    int encrypt(byte[] in, int inOfs, int inLen, byte[] out, int outOfs) {
+        checkDataLength(processed, inLen);
 
         processAAD();
-        int remainder = len % blockSize;
-        if ( remainder > 0) {
+        // 'len' stores the length to use with buffer 'in'.
+        // 'inLen' stores the length returned by the method.
+        int len = inLen;
+
+        // if there is enough data in the ibuffer and 'in' construct a block
+        // to encrypt
+
+        if (ibuffer != null && ibuffer.size() > 0 &&
+            (inLen + ibuffer.size() >= blockSize)) {
+            ArrayUtil.nullAndBoundsCheck(in, inOfs, inLen);
+            ArrayUtil.nullAndBoundsCheck(out, outOfs, inLen);
+            byte[] block = new byte[blockSize];
+            int inLenUsed = constructBlock(ibuffer.toByteArray(), in, block);
```

Comment:
constructBlock takes 'in' but not 'inOfs'? Wouldn't the data be taken from the wrong index? No test catches this, strange?

Reply:
The assumption of this whole block here seems to be that ibuffer would not contain more than a block of buffered data? If that's the case, maybe we can just use 'ibuffer' instead of allocating a local 'block' and copy the data into it?

Reply:
This is a very seldom used code path,  There was only one existing test that accesses this code and it has not offset.

constructBlock() is a generic method.  There should be no case where ibuffer is >= blocksize during this method.   For encryption the two encrypt() only put code in ibuffer.  This method is for normal update() operations.  The other encrypt is for any of the CipherCore internal buffer needs sent to GCM before doing the doFinal.  At that point, ibuffer can be >= blocksize and doFinal will handle it.  Some of this complication will go away with phase 2 of separating GCM from CipherCore

Reply:
The current impl of constructBlock() seems to have code handling ibuffer >= blocksize scenario. Could you fix that and also pass the input offset into constructBlock() for this RFE?

Reply:
I redid the whole code

Reply:
Sure, I will be waiting for the next commit then.